viamrobotics/rdk
# [RSDK-2120]/[DATA-1155] Fix flaky TestSuccesfulWrite and TestClose collector tests.

# Context
`TestSuccesfulWrite` was flaky because it used time.Sleep instead of explicit signaling. This changes it to:
- Mock time using `github.com/benbjohnson/clock`, allowing us to simulate passes of the configured interval rather than sleeping during them.
- Using a channel to signal when data has been written after an interval has passed, rather than sleeping for some period of time. 
- `.Collect()` used to return immediately.  Now it only returns once the underlying capture goroutine has been scheduled/started.
- Since we're using signaling, the configured capture intervals were decreased from ~100ms to ~2ms (the prod sleep/ticker cutoff). The tests now run significantly faster.

This isn't perfect. We still sleep at one point in the test, which I explain in a comment. As explained in that comment, I think it's a much lower risk/less flaky sleep compared to the previous ones. I'm open to suggestions here.

`TestClose` was flaky because it worked by:
- Start a collector
- Wait <amount of time that it should have written something>
- Check that it did
- Close it
- Wait <amount of time that it should have written something>
- Check that it did not
 
 Those waits were largely replaced with signaling in a very similar way to described for `TestSuccesfulWrite` . Now it:
 - Start a collector
 - Advance a mocked time by <interval>
 - Check that something was written using the <-wrote channel
 - Close it
 - Advance a mocked time by <interval>
 - Check that nothing was written using the <-wrote channel

# Testing
Ran both 1,000 times locally without a test failure with:
`go test -race --count=1000 -run TestSuccesfulWrite`
and
`go test -race --count=1000 -run TestClose`
